research/minigo/model_utils.py

```diff
@@ -0,0 +1,206 @@
+
+  Returns:
+    The model name. For example: '000000-bootstrap.index' => '000000-bootstrap'.
+  """
+  match = re.match(MODEL_NAME_REGEX, full_name)
+  if match:
+    return match.group()
+  else:
+    return None
+
+
+def get_models(models_dir):
+  """Get all models.
+
+  Args:
+    models_dir: The directory of all models.
+
+  Returns:
+    A list of model number and names sorted increasingly. For example:
+    [(13, 000013-modelname), (17, 000017-modelname), ...etc]
+  """
+  all_models = tf.gfile.Glob(os.path.join(models_dir, '*.meta'))
+  model_filenames = [os.path.basename(m) for m in all_models]
+  model_numbers_names = sorted([
+      (detect_model_num(m), detect_model_name(m))
+      for m in model_filenames])
+  return model_numbers_names
+
+
+def get_latest_model(models_dir):
+  """Find the latest model.
+
+  Args:
+    models_dir: The directory of all models.
+
+  Returns:
+    The model number and name of the latest model. For example:
+    (17, 000017-modelname)
+  """
+  models = get_models(models_dir)
+  if models is None:
+    models = [(0, '000000-bootstrap')]
+  return models[-1]
+
+
+def compute_update_ratio(weight_tensors, before_weights, after_weights):
+  """Compute the ratio of gradient norm to weight norm."""
+  deltas = [after - before for after,
+            before in zip(after_weights, before_weights)]
+  delta_norms = [numpy.linalg.norm(d.ravel()) for d in deltas]
+  weight_norms = [numpy.linalg.norm(w.ravel()) for w in before_weights]
+  ratios = [d / w for d, w in zip(delta_norms, weight_norms)]
+  all_summaries = [
+      tf.Summary.Value(tag='update_ratios/' + tensor.name, simple_value=ratio)
+      for tensor, ratio in zip(weight_tensors, ratios)]
+  return tf.Summary(value=all_summaries)
+
+
+class UpdateRatioSessionHook(tf.train.SessionRunHook):
+  """Update the ratio of gradient norm to weight norm every N steps."""
+
+  def __init__(self, working_dir, every_n_steps=100):
+    """Initializes a `UpdateRatioSessionHook`.
+
+    Args:
+      working_dir: `str`, the directory to save the update ratios.
+      every_n_steps: `int`, update the ratio every N steps.
+    """
+    self.working_dir = working_dir
+    self.every_n_steps = every_n_steps
+    self.before_weights = None
+
+  def begin(self):
+    """Called once before using the session."""
+    self.summary_writer = SummaryWriterCache.get(self.working_dir)
+    self.weight_tensors = tf.trainable_variables()
+    self.global_step = tf.train.get_or_create_global_step()
+
+  def before_run(self, run_context):
+    """Called before each call to run()."""
+    global_step = run_context.session.run(self.global_step)
+    if global_step % self.every_n_steps == 0:
+      self.before_weights = run_context.session.run(self.weight_tensors)
+
+  def after_run(self, run_context, run_values):
+    """Called after each call to run()."""
+    if self.before_weights is not None:
+      after_weights = run_context.session.run(self.weight_tensors)
```

Comment:
Why do we call session run on these instead of just returning them in run_values? Wouldn't it be more straightforward to make sure they are returned out of the step in question, rather than grabbing them here?